### PR TITLE
Send default tasks to unversioned queue when user data disabled

### PR DESCRIPTION
**Pull request by dnr**
**What changed?**
Fixes the intended behavior when user data loading is disabled by dynamic config: to drop versioned tasks and send unversioned and "default" tasks to run on the unversioned queue. This previously worked for spooled tasks but not new tasks.

Also remove frontend logs for polls when this situation occurs. (They are still logged by the matching client at Info level though.)

**Why?**
When the config is switched off, we want workflows and workers not using the versioning feature to not be affected at all. So we should continue to match tasks without an assigned version (both "unversioned" and "default") to workers not using versioning (the "unversioned queue").

For versioned tasks, all the choice have various bad tradeoffs. So far we picked the option to drop versioned tasks, which essentially blocks versioned workflows and requires manual recovery (using admin rpc `RefreshWorkflowTasks`).

**How did you test it?**
Updated some integration tests and unit tests, ran other integration tests with user data loading disabled.

**Potential risks**
Dropping versioned tasks requires awkward manual recovery

**Is hotfix candidate?**

### CODE REVIEW

File: service/matching/matching_engine.go [hunk 1]
```diff
@@ -1503,7 +1495,11 @@ func (e *matchingEngineImpl) redirectToVersionedQueueForAdd(
 	// Have to look up versioning data.
 	userData, userDataChanged, err := baseTqm.GetUserData(ctx)
 	if err != nil {
-		return taskQueue, userDataChanged, err
+		if errors.Is(err, errUserDataDisabled) && buildId == "" {
+			// When user data disabled, send "default" tasks to unversioned queue.
+			return taskQueue, nil, nil
```
Comment (bergundy) [inline service/matching/matching_engine.go:1500]: I think you'll still want to return the `userDataChanged` channel here instead of explicitly returning `nil` because user data may later be enabled (when we start periodically reading this config).

File: tests/versioning_test.go [hunk 2]
```diff
@@ -1481,6 +1481,8 @@ func (s *versioningIntegSuite) TestDisableLoadUserData() {
 	// First insert some data (we'll try to read it below)
 	s.addNewDefaultBuildId(ctx, tq, v1)
 
+	s.unloadTaskQueue(ctx, tq)
```
Comment (bergundy) [inline tests/versioning_test.go:1484]: Maybe add a comment so the reader knows why the test unloads?

File: tests/versioning_test.go [hunk 3]
```diff
@@ -1535,16 +1564,33 @@ func (s *versioningIntegSuite) TestWorkflowGetsStuckWhenDisablingLoadingUserData
...
+	// should not run on versioned worker
+	time.Sleep(2 * time.Second)
 	s.Require().Equal(int32(0), runs.Load())
+
+	wrk.Stop()
+
+	// start unversioned worker and let task run there
+	wrk2 := worker.New(s.sdkClient, tq, worker.Options{
+		BuildID:                          "v1",
+		UseBuildIDForVersioning:          false,
```
Comment (bergundy) [inline tests/versioning_test.go:1580]: Seems irrelevant

File: tests/versioning_test.go [hunk 4]
```diff
@@ -1535,16 +1564,33 @@ func (s *versioningIntegSuite) TestWorkflowGetsStuckWhenDisablingLoadingUserData
...
-		WorkflowExecutionTimeout: 5 * time.Second,
+		WorkflowExecutionTimeout: 10 * time.Second,
 	}, "wf")
 	s.Require().NoError(err)
-	err = run.Get(ctx, nil)
-	var timeoutError *temporal.TimeoutError
-	s.Require().ErrorAs(err, &timeoutError)
+
+	// should not run on versioned worker
+	time.Sleep(2 * time.Second)
```
Comment (bergundy) [inline tests/versioning_test.go:1572]: I have a couple of issues here:

1. Relying on sleeps in tests is a recipe for flakiness, I'd rather avoid that if possible.
2. More importantly, should we treat versioned workers as unversioned when user data is disabled? If we ever turn on the kill switch, it's going to be very painful for users, requiring new worker deployments.

File: service/matching/matching_engine.go (hunk 1 above)
Comment (dnr) [reply service/matching/matching_engine.go:1500]: Ah yeah, that's true

File: tests/versioning_test.go (hunk 2 above)
Comment (dnr) [reply tests/versioning_test.go:1484]: ok

File: tests/versioning_test.go (hunk 4 above)
Comment (dnr) [reply tests/versioning_test.go:1572]: 1. I mean, sure, ideally. But we're pretty far gone down the timing road in these tests. Do you have a specific suggestion for avoiding sleeps here? Note that the existing test also implicitly relied on a sleep: the 5s timeout.
2. Well, the current version of the PR does not send versioned tasks to unversioned anymore. So it's still going to be painful but in a different way (refreshing tasks). I'm open to ideas here but we discussed them all before and I'm not sure we're going to come to a different conclusion

File: tests/versioning_test.go (hunk 3 above)
Comment (dnr) [reply tests/versioning_test.go:1580]: I thought "use versioning: false" would be nice to be explicit. I can remove

File: tests/versioning_test.go (hunk 4 above)
Comment (dnr) [reply tests/versioning_test.go:1573]: I've been playing with the DLQ idea for versioned tasks and I think it's a better solution. I'll do it in a follow-up PR. This one should be merged first since it fixes an existing bug and adds more tests.

Comment (bergundy) [reply tests/versioning_test.go:1573]: > Note that the existing test also implicitly relied on a sleep: the 5s timeout.

That's not a flaky condition though, it should deterministically fail.
I don't have good idea though for an integration test that can't inspect the internal state of the system.

For (2), let's keep the decision not to break the semantics.

**Review by bergundy — APPROVED** [review]

File: tests/versioning_test.go (hunk 4 above)
Comment (dnr) [reply tests/versioning_test.go:1573]: The timeout is just as flaky as this 2s sleep: both of them can spuriously pass, but not spuriously fail. It's just the history service noticing that 5s has passed without the workflow completing yet, vs this test noticing that 2s has passed without the workflow running yet.

File: service/matching/matching_engine_test.go [hunk 5]
```diff
@@ -547,6 +547,31 @@ func (s *matchingEngineSuite) TestPollActivityTaskQueues_NamespaceHandover() {
...
+			WorkerVersionCapabilities: &commonpb.WorkerVersionCapabilities{
+				BuildId:       "some_build_id",
+				UseVersioning: true,
+			},
+		},
+	}, metrics.NoopMetricsHandler)
+	s.Nil(resp)
+	s.Error(err)
+	var failedPrecondition *serviceerror.FailedPrecondition
+	s.True(errors.As(err, &failedPrecondition))
```
Comment (MichaelSnowden) [inline service/matching/matching_engine_test.go:572]: nit: s.ErrorAs

File: service/matching/matching_engine_test.go [hunk 6]
```diff
@@ -547,6 +547,31 @@ func (s *matchingEngineSuite) TestPollActivityTaskQueues_NamespaceHandover() {
...
+			},
+			Identity: "identity",
+			WorkerVersionCapabilities: &commonpb.WorkerVersionCapabilities{
+				BuildId:       "some_build_id",
+				UseVersioning: true,
+			},
+		},
+	}, metrics.NoopMetricsHandler)
+	s.Nil(resp)
+	s.Error(err)
```
Comment (MichaelSnowden) [inline service/matching/matching_engine_test.go:570]: nit: assertions on error before result

File: service/matching/matching_engine_test.go (hunk 5 above)
Comment (dnr) [reply service/matching/matching_engine_test.go:572]: ah, nice

File: tests/versioning_test.go [hunk 7]
```diff
@@ -1506,7 +1509,34 @@ func (s *versioningIntegSuite) TestDisableLoadUserData() {
 	s.Require().ErrorAs(err, &failedPreconditionError)
 }
 
-func (s *versioningIntegSuite) TestWorkflowGetsStuckWhenDisablingLoadingUserData() {
+func (s *versioningIntegSuite) TestDisableUserData_UnversionedWorkflowRuns() {
+	tq := s.T().Name()
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
```
[the remark this answers is not in the document]
Comment (dnr) [reply tests/versioning_test.go:1515]: `context.WithTimeout`/`defer cancel()` is such a common pattern, I'd rather use it here also so it's recognized as that

File: service/frontend/workflow_handler.go [hunk 8]
```diff
@@ -878,22 +878,25 @@ func (wh *WorkflowHandler) PollWorkflowTaskQueue(ctx context.Context, request *w
...
+		switch err.(type) {
+		case *serviceerror.NewerBuildExists:
+			// expected when versioned poller is superceded
+		case *serviceerror.FailedPrecondition:
+			// server has LoadUserData switched off and this is a versioned poller
+		default:
+			// for all other errors log:
+			ctxTimeout := "not-set"
+			ctxDeadline, ok := ctx.Deadline()
+			if ok {
```
[the remark this answers is not in the document]
Comment (dnr) [reply service/frontend/workflow_handler.go]: (reverted now, but fyi this wasn't new code, I just indented it. you might want to look with whitespace diffs off, it's easier to read)

**Review by MichaelSnowden — APPROVED** [review]
